src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,191 @@
+            //corresponds to the item in delta list(Edmchangedobjectcoll). For this we use somewhat the same logic used in 
+            //the method private static PropertyConfiguration GetKeyProperty(EntityTypeConfiguration entityType) in EntityKeyConvention class
+            //Once we find the key we use that to pick the corresponding item from the original collection (by comparing the value of the key, eg: Id)
+            Type type = original.First().GetType();
+            string key = GetKeyProperty(type.GetProperties(), type.Name);
+
+            foreach (dynamic changedObj in _items)
+            {
+                object Id;
+                IEdmDeltaDeletedEntityObject deletedObj = changedObj as IEdmDeltaDeletedEntityObject;
+
+                if(deletedObj != null)
+                {
+                    TStructuralType originalObj = original.FirstOrDefault(x => x.GetType().GetProperty(key).GetValue(x).ToString() == deletedObj.Id);
+
+                    if (originalObj != null)
+                    {
+                        //This case handle deletions
+                        original.Remove(originalObj);
+                    }
+                }
+                else
+                {                 
+                    changedObj.TryGetPropertyValue(key, out Id);
+                    TStructuralType originalObj = original.FirstOrDefault(x => x.GetType().GetProperty(key).GetValue(x).ToString() == Id.ToString());
+
+                    if (originalObj == null)
+                    {
+                        //This case handle additions
+                        originalObj = Activator.CreateInstance(changedObj.ExpectedClrType);
+                        original.Add(originalObj);
+                    }
+
+                    //Patch for addition/update. This will call Delta<T> for each item in the collection
+                    changedObj.Patch(originalObj);
+                }               
+            }
+        }
+
+        /// <summary>
+        /// Patch for EdmChangedobjectCollection, a collection for Delta<typeparamref name="TStructuralType"/>
+        /// </summary>
+        /// <param name="original"></param>        
+        public void Patch(ICollection<TStructuralType> original)
+        {
+            CopyChangedValues(original);
+        }
+
+        private static string GetKeyProperty(PropertyInfo[] allProperties, string entityName)
+        {
+            var keys =
+               allProperties
+               .Where(p => (p.Name.Equals(entityName + "Id", StringComparison.OrdinalIgnoreCase) || p.Name.Equals("Id", StringComparison.OrdinalIgnoreCase))
+               && (EdmLibHelpers.GetEdmPrimitiveTypeOrNull(p.PropertyType) != null || TypeHelper.IsEnum(p.PropertyType)));
+
+            if (keys.Count() == 1)
+            {
+                return keys.Single().Name;
+            }
```

Comment:
maybe just return keys.SingleOrDefault()?.Name ?  the current logic iterates twice; once for a count and a second time for getting the value.